### PR TITLE
Feature/decl assign params

### PR DESCRIPTION
**Pull request by SteveBronder**
Adds an `assignment` member to the `Decl` record which allows for assignments made inline with the declaration to be a single line. For now this is only being done for parameters. Parameters are also given the auto type so that Eigen matrices and vectors can come in as `Eigen::Map` types. This will stop parameters from being copied when used in later functions.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Make parameters `auto` so that the parameter types for matrices can be Eigen maps

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by SteveBronder** [comment]
@WardBrian this seems to be failing because of a failed formatting, but I ran the formatter locally and seems fine? Tried to update all my stuff from opam and the scripts but I'm still getting the same formatting.

File: src/middle/Stmt.ml [hunk 1]
```diff
@@ -25,7 +25,8 @@ module Fixed = struct
           { decl_adtype: UnsizedType.autodifftype
           ; decl_id: string
           ; decl_type: 'a Type.t
-          ; initialize: bool }
+          ; initialize: bool
+          ; assignment: 'a option }
```
Comment (nhuurre) [inline src/middle/Stmt.ml:29]: `initialize` is irrelevant if `assignment` is set, right? I'd rather change this to
```suggestion
          ; initialize: 'a decl_init }
```
with
```ocaml
type 'a decl_init = Uninit | Default | Assign of 'a
```
where the first one means `initialize=false`, the second means `initialize=true` and the last one means `assignment=Some`.

**Review by nhuurre — COMMENTED** [review]
The failure is in `Debug_data_generation_tests.ml`. Some of the randomly generated numbers are off, but only in the least significant bit. Floating-point numbers are notoriously finicky. Could be some difference in rounding mode or whatever?

Running locally, I see the same results as Jenkins. You can just revert `Debug_data_generation_tests.ml`, this PR shouldn't affect that file anyway.

**Comment by SteveBronder** [comment]
> The failure is in Debug_data_generation_tests.ml. Some of the randomly generated numbers are off, but only in the least significant bit. Floating-point numbers are notoriously finicky. Could be some difference in rounding mode or whatever?

Thanks, fixed!

File: src/middle/Stmt.ml (hunk 1 above)
Comment (WardBrian) [reply src/middle/Stmt.ml:29]: I agree with the suggestion, but I wonder if it couldn't go further? The optimization detects an assignment in order to set `initialize`, so couldn't it just move the assignment inside the decl ?

Comment (SteveBronder) [reply src/middle/Stmt.ml:29]: ooh thats a pretty way to do this! I'll update

**Comment by SteveBronder** [comment]
So I absolutely need a thorough set of eyes on the optimizer code. I _think_ I got the lazy code motion correct

File: src/middle/Stmt.ml (hunk 1 above)
Comment (SteveBronder) [reply src/middle/Stmt.ml]: > I agree with the suggestion, but I wonder if it couldn't go further? The optimization detects an assignment in order to set initialize, so couldn't it just move the assignment inside the decl ?

I agree with this but think having all the decls inline will be enough changes to have a separate PR

File: src/frontend/Ast_to_Mir.ml [hunk 2]
```diff
@@ -458,7 +458,8 @@ let create_decl_with_assign decl_id declc decl_type initial_value transform
             () } in
   let decl =
     Stmt.
-      { Fixed.pattern= Decl {decl_adtype; decl_id; decl_type; initialize= true}
+      { Fixed.pattern=
+          Decl {decl_adtype; decl_id; decl_type; initialize= Default}
```
Comment (WardBrian) [inline src/frontend/Ast_to_Mir.ml:462]: I think this should use the logic on the next line which inspects the rhs_assignment to set `Initialize`

File: src/stan_math_backend/Transform_Mir.ml [hunk 3]
```diff
@@ -764,9 +764,27 @@ let add_reads vars mkread stmts =
...
   let add_read_to_decl (Stmt.Fixed.{pattern; _} as stmt) =
     match pattern with
-    | Decl {decl_id; _} when Map.mem var_names decl_id ->
+    | Decl {decl_id; decl_adtype; decl_type; _} when Map.mem var_names decl_id
+      -> (
         let loc, out = Map.find_exn var_names decl_id in
-        stmt :: mkread (Stmt.Helpers.lvariable decl_id, loc, out)
+        let param_reader = mkread (Stmt.Helpers.lvariable decl_id, loc, out) in
+        match param_reader with
+        | [ Stmt.Fixed.
```
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:772]: This looks pretty difficult to maintain and should probably be folded into mkread somehow

File: test/integration/good/code-gen/complex_numbers/cpp.expected [hunk 4]
```diff
@@ -240,43 +240,22 @@ class basic_op_param_model final : public model_base_crtp<basic_op_param_model>
     // suppress unused var warning
     (void) function__;
     try {
-      Eigen::Matrix<std::complex<local_scalar_t__>,-1,-1> cmat =
-        Eigen::Matrix<std::complex<local_scalar_t__>,-1,-1>::Constant(N, N,
-          std::complex<local_scalar_t__>(DUMMY_VAR__, DUMMY_VAR__));
-      current_statement__ = 1;
```
Comment (WardBrian) [inline test/integration/good/code-gen/complex_numbers/cpp.expected:246]: The positioning is going missing here

**Review by WardBrian — CHANGES_REQUESTED** [review]

File: src/frontend/Ast_to_Mir.ml (hunk 2 above)
Comment (SteveBronder) [reply src/frontend/Ast_to_Mir.ml:462]: Looking at this code it's not modifying the params. Making changes in places that do no modify params is out of the scope of this PR.

Comment (WardBrian) [reply src/frontend/Ast_to_Mir.ml:462]: Sounds good -- I think we would really want to fix https://github.com/stan-dev/stanc3/issues/1295 before doing this -- probably by writing a generic-ish `check_size` function in Stan which we can call after the assignment is completed.

File: test/integration/good/code-gen/complex_numbers/cpp.expected (hunk 4 above)
Comment (WardBrian) [reply test/integration/good/code-gen/complex_numbers/cpp.expected:246]: I found where they had gone

**Review by WardBrian — APPROVED** [review]
Ok @SteveBronder, assuming you don't take issue with any of the changes I just pushed this should be good to merge after Jenkins is done

**Pull request merged**